csrc/adam/cpu_adam.cpp

```diff
@@ -672,11 +672,21 @@ int ds_adam_step_plus_copy(int optimizer_id,
     return 0;
 }
 
+int destroy_adam_optimizer(int optimizer_id)
+{
+    // std::cout << "Adam Optimizer #" << optimizer_id
+    //          << " is destroyed." << std::endl;
+    s_optimizers.erase(optimizer_id);

```

Comment:
I think we need to destroy all optimizers here, when the __del__ function is called on the Python side, right? Then, I would use the .clear() rather than erase one element.

Reply:
`DeepSpeedCPUAdam.__init__` creates only one, so its corresponding ``DeepSpeedCPUAdam.__del__` should free only that one. 

It surely could delete the whole thing if it's empty to be super-clean.

Perhaps you don't need an array of optimizers in the first place if it's always one record anyway.

I just don't know the intricacies of why it was implemented this way in first place (i.e. to support multiple optimizers), so it's hard for me to know the right approach here.

Reply:
I think the way we have it currently sounds fine! thanks Stas :)